### PR TITLE
Add octree and async to preferences

### LINKED ISSUE #2712: Move NAPARI_OCTREE and NAPARI_ASYNC to settings
## 🚀 Feature
We should move the current NAPARI_OCTREE and NAPARI_ASYNC env vars to our settings. Adjusting these may require restarts.

They will probably go in an experimental/ developer section. We should also merge in the custom config stuff that was done for async into our settings

### PR DESCRIPTION
**Pull request by ppwadhwa**
# Description

Add checkboxes to the preferences menu for NAPARI_ASYNC and NAPARI_OCTREE settings.  

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

closes #2712

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/utils/settings/_defaults.py [hunk 1]
```diff
@@ -394,4 +394,39 @@ class NapariConfig:
         preferences_exclude = ['schema_version', 'disabled_plugins']
 
 
-CORE_SETTINGS = [AppearanceSettings, ApplicationSettings, PluginsSettings]
+class ExperimentalSettings(BaseNapariSettings):
+    """Developer/Experimental Settings."""
+
+    schema_version: SchemaVersion = (0, 1, 1)
+
+    NAPARI_OCTREE: bool = Field(
```
Comment (goanpeca) [inline napari/utils/settings/_defaults.py]: The name should be `octree`

File: napari/utils/settings/_defaults.py [hunk 2]
```diff
@@ -394,4 +394,39 @@ class NapariConfig:
...
+        True,
+        title=trans._("NAPARI_OCTREE"),
+        description=trans._("Use OctreeImage instead of Image."),
+    )
+
+    NAPARI_ASYNC: bool = Field(
```
Comment (goanpeca) [inline napari/utils/settings/_defaults.py]: The name should be `async_`  and we need an alias as `async` if possible

File: napari/utils/settings/_defaults.py [hunk 3]
```diff
@@ -394,4 +394,39 @@ class NapariConfig:
         preferences_exclude = ['schema_version', 'disabled_plugins']
 
 
-CORE_SETTINGS = [AppearanceSettings, ApplicationSettings, PluginsSettings]
+class ExperimentalSettings(BaseNapariSettings):
+    """Developer/Experimental Settings."""
```
Comment (goanpeca) [inline napari/utils/settings/_defaults.py]: We should remove this as we now add this on the `schema_extra` dict

File: napari/utils/settings/_defaults.py (hunk 2 above)
Comment (goanpeca) [reply napari/utils/settings/_defaults.py]: By default out settings prepend `napari_` to the model options, so currently we can use

```bash
$ NAPARI_THEME=light napari
```

That is why those names need to be just `octree` and `async`, but the latter being a reserved python word, we need a workaround

See:

https://pydantic-docs.helpmanual.io/usage/settings/

>Custom environment variable names can be set in two ways:
Config.fields['field_name']['env'] (see auth_key and redis_dsn above)
Field(..., env=...) (see api_key above)



We need to use `env="napari_async"` or `env="async"` we need to test which one works

Comment (ppwadhwa) [reply napari/utils/settings/_defaults.py]: Yeah, I see what you mean.  I guess I was thinking that for anyone who was setting this via environment variable on the widget would more likely recognize it if it looked the same as setting the environment variable.  But, if they know to change it, they will get it.

Comment (ppwadhwa) [reply napari/utils/settings/_defaults.py]: Nevermind, I see that this is not a problem anyway.  The printed name on the widget comes from the "title".  How do we want it to look on the widget?  "NAPARI_OCTREE" or "OCTREE" or "octree"?

Comment (goanpeca) [reply napari/utils/settings/_defaults.py]: >"NAPARI_OCTREE" or "OCTREE" or "octree"?

I do not think we need to show the exact env names. @sofroniewn ?

`title ="Use octree"` `description="??"`
`title ="Use async mode"` `description="??"`

Comment (sofroniewn) [reply napari/utils/settings/_defaults.py]: Yeah agreed - I might actually prefer not to even use the word octree here, it's not actually accurate to what is implemented and most users wont know what it is. I can add some suggestions here in the morning. Maybe @jni has ideas too

Comment (ppwadhwa) [reply napari/utils/settings/_defaults.py]: @sofroniewn @jni 

I made an attempt at descriptive names.  What do you think?

Screen Shot 2021-05-27 at 1 40 36 PM

**Comment by ppwadhwa** [comment]
I'm not sure why the pre-commit hooks are failing (I just ran them), but will check this in a bit.

File: napari/utils/settings/_defaults.py [hunk 4]
```diff
@@ -394,4 +394,46 @@ class NapariConfig:
...
+class ExperimentalSettings(BaseNapariSettings):
+
+    schema_version: SchemaVersion = (0, 1, 1)
+
+    octree: bool = Field(
+        True,
+        title=trans._("Enable Tiling of Images"),
+        description=trans._(
+            "Renders images using tiles. \nYou must restart napari for "
+            + "changes of this setting to apply."
```
Comment (sofroniewn) [inline napari/utils/settings/_defaults.py]:
```suggestion
        title=trans._("Enable Asynchronous Tiling of Images"),
        description=trans._(
            "Renders images asynchronously using tiles. \nYou must restart napari for "
            + "changes of this setting to apply."
```

File: napari/utils/settings/_defaults.py (hunk 2 above)
Comment (sofroniewn) [reply napari/utils/settings/_defaults.py]: Loving it! I suggest that we add `asynchronous` in front of `tiling` as well, to make it clear that it is asynchronous tiles, we can refine further later - but this looks like a great start!

Comment (goanpeca) [reply napari/utils/settings/_defaults.py]: They should be False by default, right @sofroniewn ?

**Comment by goanpeca** [comment]
@sofroniewn are these two options intertwined? I am looking at the code and see:

```python
def _get_async_config() -> Optional[dict]:
    # ....
    if async_var:
        async_config = DEFAULT_OCTREE_CONFIG.copy()
        async_config['octree']['enabled'] = False
        return async_config


def get_octree_config() -> dict:
     # ...
    if octree_var in [None, False]:
        # This will return DEFAULT_ASYNC_CONFIG or None.
        return _get_async_config()

    # If NAPARI_OCTREE is "1" then use default config.
    if octree_var == "1":
        return DEFAULT_OCTREE_CONFIG
    # ...
```

So if both are True, it seem it is only as if OCTREE is On and Async is Off?

Or to put it another Way

1. `async = False` and `Octree = False` == Nothing is used
2. `async = True` and `Octree = False` == Async is used
3. `async = True` and `Octree = True` == Octree is used
4. `async = False` and `Octree = True` == Octree is used

So 3 and 4 are produce the same result?

File: napari/utils/settings/_defaults.py (hunk 2 above)
Comment (sofroniewn) [reply napari/utils/settings/_defaults.py]: all these should be false by default

**Comment by sofroniewn** [comment]
> So if both are True, it seem it is only as if OCTREE is On and Async is Off?

Yes that is the correct logic, it's a bit funny because the OCTREE has to be asynchronous but it is different than ASYNC, and we made it take precedence.

Very open to suggestions to how we might want to communicate that to a user. Maybe if OCTREE is True then we could disable/ gray out Async to indicate it has no effect?

I think with time we'll clean this up more significantly / make some larger changes to how this is organized, controlled. But this will at least let a GUI user try this stuff out

**Comment by goanpeca** [comment]
>Very open to suggestions to how we might want to communicate that to a user. Maybe if OCTREE is True then we could disable/ gray out Async to indicate it has no effect?

Yep, I like that one

**Comment by ppwadhwa** [comment]
https://user-images.githubusercontent.com/54282105/120267884-775f7a80-c26a-11eb-8d8d-b1dd78d2da26.mov

**Comment by sofroniewn** [comment]
This looks like it is coming along nicely - it looks like the async option is disabled when the tiled is selected (i can't change it's selection) but can we also gray it out so i'm not tempted to click on it. this is what i see now 
image

i think once that is done this will be good to merge. main functionality worked as expected for me!!

**Comment by goanpeca** [comment]
>This looks like it is coming along nicely - it looks like the async option is disabled when the tiled is selected (i can't change it's selection) but can we also gray it out so i'm not tempted to click on it. this is what i see now

It seems we do not have the correct styling for disabled checkboxes. But yes agreed!

**Comment by ppwadhwa** [comment]
@sofroniewn @goanpeca I have made changes so that the checkbox for async dims when octree is selected.  

I also added in some code to dim widgets that have been disabled in preferences due to having an environment variable already set for it ( disabled widgets happened in #2797).   

Let me know what you think!  thanks!


https://user-images.githubusercontent.com/54282105/120411617-b5b47280-c31a-11eb-9cfc-8f4a1a9a3a2a.mov

**Review by sofroniewn — APPROVED** [review]
Disabled behaviour in the video looks great to me. Thanks!!

**Pull request merged**